### PR TITLE
Feature/16402 handle refresh during check

### PR DESCRIPTION
**Pull request by jon-shipley**
Handle page refreshes during the check by reading and initialising current state

* update inputService to save to storage immediately rather than on flush
* refreshing on the warmup or real questions cause the user to be moved to the next state, other views are allowed to refresh

### CODE REVIEW

File: pupil-spa/src/app/check/check.component.spec.ts [hunk 1]
```diff
@@ -236,4 +248,140 @@ describe('CheckComponent', () => {
...
+      // set up test spy on Audit Service
+      component.refreshDetected()
+      // check the spy was called
+      expect(auditService.addEntry).toHaveBeenCalledTimes(1)
+      expect(auditEntryInserted instanceof RefreshDetected).toBeTruthy()
+    })
+
+    it('moves to the next state if a real test question was asked', () => {
+      // find the state for the first warmup question
+      let q1 = component['allowedStates'].indexOf('Q1');
```
Comment (GeorgeChatzigiannis) [inline pupil-spa/src/app/check/check.component.spec.ts]: const

File: pupil-spa/src/app/check/check.component.spec.ts [hunk 2]
```diff
@@ -236,4 +248,140 @@ describe('CheckComponent', () => {
...
+      // exercise the code
+      component.ngOnInit()
+
+      // test
+      expect(component['state']).toBe(q1 + 1)
+    })
+
+    it('state stays the same if a page refresh happens on a warmup loading screen', () => {
+      // find the state for the first warmup question
+      let state = component['allowedStates'].indexOf('LW1');
```
Comment (GeorgeChatzigiannis) [inline pupil-spa/src/app/check/check.component.spec.ts]: state is not modified (const)

File: pupil-spa/src/app/check/check.component.ts [hunk 3]
```diff
@@ -27,6 +27,14 @@ export class CheckComponent implements OnInit {
   public allowedStates: Array<string> = [];
   private totalNumberOfQuestions: number;
   public static readonly checkStateKey = 'checkstate';
+  private static warmupIntroRe = /^warmup-intro$/;
```
Comment (GeorgeChatzigiannis) [inline pupil-spa/src/app/check/check.component.ts]: Should we have a statics file for these?
Just so to keep purely the functional part for the service here.

Comment (jon-shipley) [reply pupil-spa/src/app/check/check.component.ts]: The state is internal to the component, so it makes sense to keep it together to me.

**Comment by GeorgeChatzigiannis** [comment]
is this stilll WIP as stated in the description or is it 100% rdy for review?

**Comment by jon-shipley** [comment]
It's WIP still

File: pupil-spa/src/app/check/check.component.ts [hunk 4]
```diff
@@ -234,4 +261,81 @@ export class CheckComponent implements OnInit {
...
+    this.auditService.addEntry(new RefreshDetected());
+
+    // Lets say that handling reloads during the check should always show the current screen
+    // in which case handling the reload whilst a question was being shown is a special case.
+    if (CheckComponent.questionRe.test(stateDesc)) {
+      // the page was reloaded when a question was shown
+      // Store the answer as the empty string (as there was no input)
+      // we need to initialise the current question, with the one from the current state
+      const matches = CheckComponent.questionRe.exec(stateDesc);
+      const questionNum = parseInt(matches[ 1 ], 10);
```
Comment (GeorgeChatzigiannis) [inline pupil-spa/src/app/check/check.component.ts]: isn't radix (second arg) defaulting to 10 anyway?

Comment (jon-shipley) [reply pupil-spa/src/app/check/check.component.ts]: The recommendation is always use it, so I do :)
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

File: pupil-spa/src/app/check/check.component.ts [hunk 5]
```diff
@@ -80,10 +84,29 @@ export class CheckComponent implements OnInit {
...
-    this.state = 0;
-    this.isWarmUp = true;
-    this.viewState = 'warmup-intro';
-    this.totalNumberOfQuestions = this.warmupQuestionService.getNumberOfQuestions();
+    // set up the state
+    if (this.storageService.getItem(CheckComponent.checkStateKey)) {
+      // existing check state detected
+      // assume we are reloading during a check
+      const existingState = this.storageService.getItem(CheckComponent.checkStateKey);
+      if (!this.isValidState(existingState)) {
```
Comment (GeorgeChatzigiannis) [inline pupil-spa/src/app/check/check.component.ts]: @jon-shipley not a big fan of nested if as they oppose the principles of clean coding.
How about you approach it with a gateway method?
https://wpshout.com/unconditionally-refactoring-nested-statements-cleaner-code/

Comment (GeorgeChatzigiannis) [reply pupil-spa/src/app/check/check.component.ts]: if (!this.storageService.getItem(CheckComponent.checkStateKey)) {
..code 
return
}

Comment (jon-shipley) [reply pupil-spa/src/app/check/check.component.ts]: @GeorgeChatzigiannis  I have refactored the code block to make it clearer.

**Pull request merged**